spacy/language.py

```diff
@@ -1988,11 +2016,52 @@ def to_disk(
         serializers["vocab"] = lambda p: self.vocab.to_disk(p, exclude=exclude)
         util.to_disk(path, serializers, exclude)
 
+    @staticmethod
+    def _resolve_component_activation_status(
+        inactive: Iterable[str],
+        active: Iterable[str],
+        component_names: Collection[str],
+        arg_type: str,
+    ) -> Iterable[str]:
+        """
+        Merges arguments in spacy.load(), lang_cls.from_config() etc. reflecting components' activation status.
+        `inactive` can be the values passed as `excluded` or `disabled`, `active` the values of `included` or `enabled`.
+        The resolution logic is the same: `active` takes precedence over `inactive` - so if the former is set, all other
+        components are assumed to be disabled, independently from the value of `inactive`. If `active` is empty, only
+        those component whose names are included in `inactive` are assumed to be disabled.
+
+        inactive (Iterable[str]): Names of inactive components or serialization fields (i.e. to be excluded or
+                                  disabled).
+        active (Iterable[str]): Names of active pipeline components (i.e. to be included or enabled).
+        component_names (Iterable[str]): Names of all pipeline components.
+        arg_type (str): One of (exclude, disable). Used only to generate fitting warning message in case of possibly
+                                                   ambiguous values for `inactive`/`active`.
+
+        RETURNS (Collection[str]): Names of components to exclude from pipeline w.r.t. specified includes and excludes.
+        """
+
+        assert arg_type in ("exclude", "disable")
+
+        if active:
+            if inactive:
+                warnings.warn(
+                    Warnings.W120.format(
+                        arg1="include" if arg_type == "exclude" else "enable",
+                        arg2=arg_type,
+                    )
+                )
```

Comment:
This is a reasonable strategy, but I wonder whether it would be confusing because we do something slightly different in `nlp.select_pipes`: There we raise an exception if the two settings are both set but notcompatible:
```
        if enable is not None:  
            ...
            to_disable = [pipe for pipe in self.pipe_names if pipe not in enable]
            
            # raise an error if the enable and disable keywords are not consistent
            if disable is not None and disable != to_disable:
                raise ValueError(
                    Errors.E992.format(
                        enable=enable, disable=disable, names=self.pipe_names
                    )
                )
```

Reply:
Hm, I didn't consider that. I agree that we should keep the warnings consistent. Then I'll scrap `W120` and go with `E992` on incompatibility?

Reply:
Using a similar mechanism and error message now - let me know what you think. Latest commit also includes a minor test refactoring. Apparently I also broke something, since `test_load_without_strings()` fails now. Will take a look at that next week.

And sorry for rebasing - force of habit, noticed too late :grimacing: Last time :crossed_fingers:

Reply:
There's not a complete list of fields supported by `exclude` for serialization and it's allowed for components to define their own, although I doubt it's done much in practice. `exclude` is passed to all the serialization methods so this needs to be supported.

I think you have to allow arbitrary extra things in `exclude`.

Reply:
Then I think the contrast between `exclude` (anything) and `include` (pipeline components) is going to be weird. How weird would it be to just allow extra ignored things in `include` like with `exclude`?

Reply:
Although I guess `disable` is also (pipeline components), so maybe it's not that weird. But the checks are going to need to allow arbitrary things in `exclude`.

Reply:
I quite dislike this inconsistency, I think it's rather confusing from a usage perspective. I don't see how we can avoid it however as long as it's not possible to retrieve all relevant fields.

I relaxed the check now so that arbitrary terms are allowed in `exclude`, but ignored once `include` is not empty. This leads to inconsistent component exclusions:
```python
# "strings" is not in nlp.pipe_names

# Will exclude ("strings",)
spacy.load(tmp_dir, exclude=["strings"])  
# Will exclude all components in pipeline except "tagger", but not "strings"
spacy.load(tmp_dir, include=["tagger"], exclude=["strings"])  
```
I believe this to be an rather unexpected result. This could be handled differently by adding arbitrary elements in `exclude` (such as `"strings"`) to the exclusions even if `included` is set. This would be inconsistent in different ways though:
```python
# "strings" is not in nlp.pipe_names

# Will exclude "strings" + all components in pipeline except "tagger" 
spacy.load(tmp_dir, include=["tagger"], exclude=["strings"]) 
# Will exclude all components in pipeline "except" tagger, but not "strings" (as we are not aware of it)
spacy.load(tmp_dir, include=["tagger"])  
```
Again not ideal IMO, as I'd expect `include` to exclude _everything_ else.

If I understand correctly, option (1) is more closely aligned with the behaviour of `select_pipes()`, so I'd keep this and adjust the documentation, if we agree on this. Perhaps creating an issue for looking into deeper into this and fixing the `include`/`exclude` behaviour to be consistent would be reasonable?